### ISSUE TITLE
Attribute width not allowed on element source at this point.

### ISSUE THREAD

**Issue by serg-quinta** [validator/validator#1161]
Hello!

I've added width and height attributes to the `source` element according next specification:

https://html.spec.whatwg.org/multipage/embedded-content.html#the-source-element

```
Content attributes:
Global attributes
type — Type of embedded resource
src (in video or audio) — Address of the resource
srcset (in picture) — Images to use in different situations, e.g., high-resolution displays, small monitors, etc.
sizes (in picture) — Image sizes for different page layouts
media (in picture) — Applicable media
width (in picture) — Horizontal dimension
height (in picture) — Vertical dimension
```

and https://html.spec.whatwg.org/multipage/embedded-content-other.html#attr-dim-width

```
Author requirements: The width and height attributes on img, iframe, embed, object, video,
source when the parent is a picture element and, when their type attribute is in the Image Button state, 
input elements may be specified to give the dimensions of the visual content of the element
(the width and height respectively, relative to the nominal direction of the output medium), in CSS pixels.
The attributes, if specified, must have values that are valid non-negative integers.
```

But I've got an error.

Code to reproduce error:

```
    <picture>↩
        <source srcset="small-img.jpg"↩
                media="(max-width: 739.98px)"↩
                width="320"↩
                height="320">↩
        <img src="image.jpg"↩
             alt="Image" ↩
             width="160"↩
             height="600">↩
    </picture>↩
```

Error:

```
Attribute width not allowed on element source at this point.
```

Can you check it, please?

**Comment by tannerdolby** [comment]
@serg-quinta I ran the HTML through [the validator](https://validator.w3.org/nu/#textarea) and the error you speak of is thrown as expected because `width` is not a valid attribute for `<source>` elements.

<img width="1280" alt="Screen Shot 2021-07-08 at 4 22 51 PM" src="https://user-images.githubusercontent.com/48612525/125001991-c4bccd80-e008-11eb-94a8-f763055e0e7f.png">

The valid attributes for [`<source>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source) elements don't include a `width` attr. They only expect `src`, `srcSet`, `type`, `media` etc described in [`<source>` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source#attribute)

After removing the invalid attributes in `<source>`, adding a closing `>` tag for the source element, no errors or warnings are present and the validator only complains about the "No space between attributes" because of the newline characters you've included.

<img width="1259" alt="Screen Shot 2021-07-08 at 4 34 36 PM" src="https://user-images.githubusercontent.com/48612525/125002722-68f34400-e00a-11eb-9020-95b309c3ba3a.png">

**Comment by tannerdolby** [comment]
Once the OP removes `width` and `height` attributes from the `<source>` element, the errors mentioned in issue will be fixed.

I reckon this can be closed whenever @sideshowbarker !

**Comment by sideshowbarker** [comment]
Thanks @tannerdolby — closing per https://github.com/validator/validator/issues/1161#issuecomment-876809552

**Comment by tannerdolby** [comment]
Your welcome @sideshowbarker !

**Comment by tannerdolby** [comment]
Your welcome @sideshowbarker ! Glad to help.

**Comment by serg-quinta** [comment]
> The [Spec](https://html.spec.whatwg.org/multipage/embedded-content.html#the-source-element) does state the `width` and `height` attributes can be used, so this is a bit confusing. It seems that the `<source>` elements place in the DOM or what its parent element is, from the "at this point" part of the error is the underlying problem.
> 
> What do you think @sideshowbarker ?

Yes, the spec does state the width and height attributes can be used. And it works in Chrome browser but fails in Safari and Firefox browsers.

**Comment by serg-quinta** [comment]
My case is that I'm trying to add different images for different screens and if I adding `width` and `height` to `img` element, image from `source` has wrong dimensions. This why I added `width` and `height` to `img` as it is allowed in [spec](https://html.spec.whatwg.org/multipage/embedded-content.html#the-source-element). 
But it hasn't cross browser support.

**Comment by tannerdolby** [comment]
> My case is that I'm trying to add different images for different screens and if I adding width and height to img element, image from source has wrong dimensions. 

First off, have a look at [serving responsive images](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images). You want to use `srcSet` and `sizes` to achieve this sort of responsive image. Using `width` and `height` are not the correct attributes in order to serve different size/pixel ratio images at specific screen sizes. 

Use `srcSet` instead and provide a `sizes` attribute which will be essentially the same as your `media` attribute. Since cross browser support isn't fully there for `width/height` on `<source>`, I'd suggest something like this:

```html
<picture>
        <source srcset="small-img.jpg 320w, med-img.jpg 640w, large-img.jpg 1024w"
                sizes="(max-width: 739.98px) 33.3vw, 100vw"
        <img src="image.jpg"
             alt="Image" 
             width="160"
             height="600">
    </picture>
```

**Comment by serg-quinta** [comment]
> ```
> srcset
> ```

`srcset` isn't my case. I need different images for different screens. This is why I used `picture` element.

**Comment by serg-quinta** [comment]
I've user `width` and `height` attributes because Google Web Vitals report me an error if these attributes is skip.

**Comment by AndySky21** [comment]
Sure thing, this error should not be flagged as "attribute 'width' not
allowed on element 'source' at this point" in the first place. The "source"
element allows for that if it is in a "picture" element, which is the case
here, and it would lead to confusion with regard to standards.

Il Lun 12 Lug 2021, 15:06 Serhiy Valchuk ***@***.***> ha
scritto:

**Comment by tannerdolby** [comment]
> srcset isn't my case. I need different images for different screens. This is why I used picture element.

Using [srcset](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture#the_srcset_attribute) with `<picture>` is exactly how you serve "images that work well on devices with widely differing screen sizes, resolutions, and other such features". Using `width` and `height` are absolute values and won't be responsive for different viewports. Is that more clear now? If you had a look at [`<picutre>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture) that should be pretty clear.

**Comment by tannerdolby** [comment]
> The "source" element allows for that if it is in a "picture" element, which is the case here, and it would lead to confusion with regard to standards.

Yep I agree, I think that's what I had a hard time coming to terms with earlier in this discussion. So I think the validator needs to recognize that the `<source>` element is nested within `<picture>` which is indeed valid HTML based on the [Spec](https://html.spec.whatwg.org/multipage/embedded-content.html#the-source-element).

I think this might need to be reopened and looked at @sideshowbarker !

**Comment by sideshowbarker** [comment]
Sorry all, I can see now that when I closed this, I was just confused.

In fact the problem is that the checker currently has no support at all for recognizing `source[width]` as valid. So I simply need to update the checker code to bring it into conformance with the spec.

**Comment by tannerdolby** [comment]
I'm happy to help bring this into conformance with spec, would adding `source.picture.attrs.width` and same for `height` to `embed.rnc` help speed things along? I saw places in Schema where `<source>` attributes were being defined, not entirely sure if the `.rnc` files contribute to what the checker expects.

`embed.rnc`
```rnc
## Picture source: <source srcset>

	source.picture.elem =
		element source { source.picture.inner & source.picture.attrs }
	source.picture.attrs =
		(	common.attrs
		&	source.picture.attrs.media?
		&	source.picture.attrs.srcset
		&	source.picture.attrs.sizes?
		&	source.picture.attrs.type?
... (additions)
                &	source.picture.attrs.width?
                &	source.picture.attrs.height? 
		)
```
